### DISCUSSION TITLE
Using encryption on targets cache

### DISCUSSION THREAD

**Discussion by ghost** [ropensci/targets#448] [Help]
Hello, 

for my analyses I often need to extract (i.e. safe) queries from a database that take quite some time. Ideal for targets to chunk them and selectively re-run the parts that changed. Only problem is, the data is extremely sensitive and having parts stored unencrypted (even on the companies file system) is unwise / untenable. 

Therefore, I use the cyphr package to encrypt/decrypt all files during the projects runtime. It usually looks like this:

```
encrypt(read_rds(<object_to_encrypt>, path = <path_to_file>, key = <encryption_key>)
decrypt(read_rds(file = <file.path>, key = <encryption_key>)
```

Is / would it be possible to define, or redefine, a `targets` storage option which incorporates this wrapper, thus encrypting and decrypting in the `targets` framework?

Thank you very much!

**Comment by wlandau** [comment]
This question is actually quite timely, I am planning a feature in `tarchetypes` that should handle this use case. Please see the encryption example at https://github.com/ropensci/tarchetypes/issues/44. Related: https://twitter.com/dirk_sch/status/1388600198738362368.

**Reply by ghost** [reply]
Thanks, I took a stab at trying to implement encryption using `tarchetypes::hook_inner/outer`. I couldnt get it to work. 
As far as I understand the hooks allow to wrap the targets functions that create the individual targets. The issue with `cyphr` encryption is that it enrypts data on write/read proccesses. So a hook or a hint would be needed so I can write a wrapper for `tar_target`s utility to save the targets. 

I looked at flow diagrams of `tar_target`, `tar_target_raw` and `target_ini`. Am I correct in assuming that a different process controls saving the data? 

I would prefer it if it was possible to re(define) a tar_target function that has a modified function for saving so that I can implement the encryption.

**Reply by wlandau** [reply]
What about `cyphr::encrypt_object()` and `cyphr::decrypt_object()` in the hooks? Those functions work on the in-memory objects themselves rather than read/write processes. The existing way `targets` saves data should be compatible with that.

> I looked at flow diagrams of tar_target, tar_target_raw and target_ini. Am I correct in assuming that a different process controls saving the data?

Saving data happens in the same R session/process that runs the pipeline except if you are using parallel workers (e.g. `tar_make_clustermq()`) and the `storage` argument is set to `"worker"`.

**Reply by ghost** [reply]
Thanks, it works using the hooks like this with the default storage option:
```
...
%>%
tar_hook_outer(hook = encrypt_object(.x, key = cyphr_key), names = c("data")) %>%
tar_hook_inner(hook = decrypt_object(.x, key = cyphr_key))
```

Now i got another curveball (sorry), when specifying the `parquet` format for storage I get the error 
`Target with format = "parquet" must return data frames.`
 The format is preferred since it allows efficient storage of large datasets. Due to infuriating hard- and software constraints at my company I cant just push the entire data set to an efficient database. The result is that I need to contort many solutions to work within these constraints.

### Options
1. Wrap the function that saves the output of targets with encrypt/decrypt
2. Using hooks to encrypt the file *after* it has been written / encrypt it *before* targets loads the file using `cyphr::decrypt_file`. 
  (2) Would involve inserting the code before load and write and getting the path targets is calling for the target in question.
3. Redefine the command/method used to save as parquet by wrapping it with cyphr::encrypt/decrypt. 

If you could point me the place I need to modify ill happily do it, sorry that I cant find it myself.



### Targets pipeline

```
library(targets)
library(tarchetypes)
 require(tidyverse)
 require(here)
 require(cyphr)
 require(keyring)


# generate sodium key
cyphr_key  <- cyphr::key_sodium(sodium::keygen())


# define targets functions
make_test <- function(){
  x <- as_tibble(iris)
  return(x)
}

summ <- function(dataset) {
  sample <-
    dataset %>% janitor::clean_names() %>%
    group_by(species) %>%
    summarise(across(.fns = list(median = median)))

  return(sample)
}



# targets -----------------------------------------------------------------


# Set target-specific options such as packages.
tar_option_set(packages = c("tidyverse", "arrow", "cyphr", "janitor"))

# End this file with a list of target objects.
#targets <-
  list(
  tar_target(data, make_test(), format = "parquet"),
  tar_target(summary, summ(data), format = "parquet") # Call your custom functions as needed
  ) %>%
tar_hook_outer(hook = encrypt_object(.x, key = cyphr_key), names = c("data")) %>%
tar_hook_inner(hook = decrypt_object(.x, key = cyphr_key))

````

**Comment by wlandau** [comment]
If you upgrade your installation of `targets` to the latest CRAN version, parquet format will work with objects of class data.frame, RecordBatch, or Table. These are all the classes that `arrow::write_parquet()` supports (according to the help file).

**Reply by ghost** [reply]
When using `encrypt_object()` you get a raw vector. Therefore you cannot store it using parquet since its not a data.frame or any data structure that is supported. 

It would work if it would encrypt the contents of the data.frame but not the entire object. Therefore the need to wrap the save_parquet with encrypt().

**Reply by wlandau** [reply]
Yeah, the encryption hook should work if format = “qs”. qs is an efficient alternative to parquet.
